src/cast.c

```diff
@@ -0,0 +1,271 @@
+#include "vctrs.h"
+
+// Initialised at load time
+static SEXP vec_cast_dispatch_fn = NULL;
+
+
+static SEXP int_as_logical(SEXP x, bool* lossy) {
+  int* data = INTEGER(x);
+  R_len_t n = Rf_length(x);
+
+  SEXP out = PROTECT(Rf_allocVector(LGLSXP, n));
+  int* out_data = LOGICAL(out);
+
+  for (R_len_t i = 0; i < n; ++i, ++data) {
+    int elt = *data;
+
+    if (elt != 0 && elt != 1) {
+      *lossy = true;
+      UNPROTECT(1);
+      return R_NilValue;
```

Comment:
Why do you have an early return here?

Reply:
Oooooh because it falls back to the regular method for lossy casts. I wonder if there's some other way to implement this that doesn't require so many contortions of the C code.

Reply:
We could wrap the dispatch function in a function and call it from the coercers. It seems cleaner to me to do the early return to make it clear that the output is produced by another mechanism though (R implementation).

This is all temporary, until we're able to signal the typed conditions from C. It requires a bit of effort because of the details on lossy casts included in the condition. I just realised the growable stuff might be helpful here.

Reply:
Ok makes sense.